### DISCUSSION TITLE
[Bug]: How to handle block function

### DISCUSSION THREAD

**Discussion by zacard-orc** [theos/theos#655] [Questions] [answered]
### What are the steps to reproduce this issue?

all the step build no error

### What happens?

can't catch wrap of block function.

when i use this code
```text
typedef void(^ozrsp)(id rsp);

%hook xxxx
- (id)requesurl:(id)arg1 data:(id)arg2 succ:(id)arg3{
void(^rspblock)(id) = ^(id xsp) {
    NSLog(@"lin rsp = %@",xsp);
	if(xsp){
		return arg3(xsp);
	} else{
		return arg3();
	}
};
return %orig(arg1,arg2,rspblock);
}

%end
```

and then only get this result when using xcode console nslog output
```text
default	11:10:55.144596+0800	OZ	lin url = https://i.bzz.com/service/settings/v2/
default	11:10:55.144614+0800	OZ	lin cb = <__NSMallocBlock__: 0x28383d350>
default	11:10:55.195818+0800	OZ	lin rsp = (null)
```
but when i using frida with the similar syntex it can get callback response

### What were you expecting to happen?

I hope to catch the origin response in block function

### What OS are you seeing the problem on?

macOS

### What OS version does it have installed?

12

### What toolchain and version are you using?

xcode13

### Which SDK version are you using?

11

### What OS is your client device running?

iOS

### What OS version does it have installed?

13

### Which device model is it?

iphone7

**Comment by kirb** [comment, marked as answer]
You may need to copy the block. By default a block is NSStackBlock, which as you can imagine, is expected to only live for a short period of time entirely on the stack. Copying will give you an NSMallocBlock that can live in the heap. As the original block is an NSMallocBlock, you likely  need to do that to match. Blocks respond to -copy, so you can try:

```logos
return %orig(arg1, arg2, [respblock copy]);
```

**Reply by zacard-orc** [reply]
thx, I got it
I retry the code and recv the right response 😃
